### PR TITLE
test: add integration test for allow-metric-label

### LINKED ISSUE #126526: allow-metric-labels can not filter metrics
### What happened?

```
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  labels:
    component: kube-scheduler
    tier: control-plane
  name: kube-scheduler
  namespace: kube-system
spec:
  containers:
  - command:
    - kube-scheduler
    - --bind-address=0.0.0.0
    - --config=/etc/kubernetes/scheduler-config.yaml
    - --v=2
    - --authentication-kubeconfig=/etc/kubernetes/scheduler.conf
    - --authorization-always-allow-paths=/healthz,/readyz,/livez,/configz,/metrics
    - --authorization-kubeconfig=/etc/kubernetes/scheduler.conf
    - --allow-metric-labels="kubernetes_feature_enabled,stage='BETA'","kubernetes_feature_enabled,name='test'"
    image: registry.aliyuncs.com/google_containers/kube-scheduler:v1.30.0
    imagePullPolicy: IfNotPresent
    livenessProbe:
      failureThreshold: 8
      httpGet:
        host: 127.0.0.1
        path: /healthz
        port: 10259
        scheme: HTTPS
      initialDelaySeconds: 10
      periodSeconds: 10
      timeoutSeconds: 15
    name: kube-scheduler
    resources:
      requests:
        cpu: 100m
    startupProbe:
      failureThreshold: 24
      httpGet:
        host: 127.0.0.1
        path: /healthz
        port: 10259
        scheme: HTTPS
      initialDelaySeconds: 10
      periodSeconds: 10
      timeoutSeconds: 15
    volumeMounts:
    - mountPath: /etc/kubernetes/scheduler.conf
      name: kubeconfig
      readOnly: true
    - mountPath: /etc/kubernetes/scheduler-config.yaml
      name: scheduler-config
      readOnly: true
    - mountPath: /etc/kubernetes/extender/
      name: cert
      readOnly: true
    - mountPath: /etc/kubernetes/certs/
      name: scheduler-cert
      readOnly: true     
  hostNetwork: true
  priority: 2000001000
  priorityClassName: system-node-critical
  securityContext:
    seccompProfile:
      type: RuntimeDefault
  volumes:
  - hostPath:
      path: /etc/kubernetes/scheduler.conf
      type: FileOrCreate
    name: kubeconfig
  - hostPath:
      path: /etc/kubernetes/scheduler-config.yaml
      type: FileOrCreate
    name: scheduler-config
  - hostPath:
      path: /etc/kubernetes/extender
      type: DirectoryOrCreate
    name: cert
  - hostPath:
      path: /etc/kubernetes/certs
      type: DirectoryOrCreate
    name: scheduler-cert
---


bernetes_feature_enabled{name="LogarithmicScaleDown",stage="BETA"} 1
kubernetes_feature_enabled{name="LoggingAlphaOptions",stage="ALPHA"} 0
kubernetes_feature_enabled{name="LoggingBetaOptions",stage="BETA"} 1
kubernetes_feature_enabled{name="MatchLabelKeysInPodAffinity",stage="ALPHA"} 0
kubernetes_feature_enabled{name="MatchLabelKeysInPodTopologySpread",stage="BETA"} 1
kubernetes_feature_enabled{name="MaxUnavailableStatefulSet",stage="ALPHA"} 0
kubernetes_feature_enabled{name="MemoryManager",stage="BETA"} 1
kubernetes_feature_enabled{name="MemoryQoS",stage="ALPHA"} 0
kubernetes_feature_enabled{name="MinDomainsInPodTopologySpread",stage=""} 1
kubernetes_feature_enabled{name="MultiCIDRServiceAllocator",stage="ALPHA"} 0
kubernetes_feature_enabled{name="MutatingAdmissionPolicy",stage="ALPHA"} 0
kubernetes_feature_enabled{name="NFTablesProxyMode",stage="ALPHA"} 0
kubernetes_feature_enabled{name="NewVolumeManagerReconstruction",stage=""} 1
kubernetes_feature_enabled{name="NodeInclusionPolicyInPodTopologySpread",stage="BETA"} 1
kubernetes_feature_enabled{name="NodeLogQuery",stage="BETA"} 0
kubernetes_feature_enabled{name="NodeOutOfServiceVolumeDetach",stage=""} 1
kubernetes_feature_enabled{name="NodeSwap",stage="BETA"} 1
kubernetes_feature_enabled{name="OpenAPIEnums",stage="BETA"} 1
kubernetes_feature_enabled{name="PDBUnhealthyPodEvictionPolicy",stage="BETA"} 1
kubernetes_feature_enabled{name="PersistentVolumeLastPhaseTransitionTime",stage="BETA"} 1
kubernetes_feature_enabled{name="PodAndContainerStatsFromCRI",stage="ALPHA"} 0
kubernetes_feature_enabled{name="PodDeletionCost",stage="BETA"} 1
kubernetes_feature_enabled{name="PodDisruptionConditions",stage="
```

### What did you expect to happen?

filter metrics

### How can we reproduce it (as minimally and precisely as possible)?

change kube-scheduler cmd line

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

1.30.0

</details>

### Cloud provider

<details>
wmware workstation
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by yongruilin**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
* Expose the ResetLabelAllowList interface for the targeted tested metrics.
* Add integration test for `--allow-metric-labels`
#### Which issue(s) this PR fixes:

Fixes #126526 

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/integration/metrics/metrics_test.go [hunk 1]
```diff
@@ -261,6 +263,90 @@ func TestAPIServerMetricsLabels(t *testing.T) {
...
+		{
+			// Histogram metric
+			metricName:  "apiserver_request_duration_seconds",
+			labelName:   "verb",
+			allowValues: model.LabelValues{"POST", "LIST"},
+			isHistogram: true,
+		},
+	}
+
+	// Assemble the allow-metric-labels flag.
```
Comment (richabanker) [inline test/integration/metrics/metrics_test.go:295]: Worth extracting the flag construction into a separate function?

File: test/integration/metrics/metrics_test.go [hunk 2]
```diff
@@ -261,6 +263,90 @@ func TestAPIServerMetricsLabels(t *testing.T) {
...
+		if allowMetric.isHistogram {
+			metricName = metricName + "_sum"
+		}
+		if samples, found := metrics[metricName]; !found {
+			t.Errorf("metric %q not found", metricName)
+		} else {
+			for _, sample := range samples {
+				if value, ok := sample.Metric[allowMetric.labelName]; ok {
+					if !slices.Contains(allowMetric.allowValues, value) && value != "unexpected" {
+						t.Errorf("value %q is not allowed for label %q", value, allowMetric.labelName)
```
Comment (richabanker) [inline test/integration/metrics/metrics_test.go:329]: should this be t.Fatalf() ? Same on L#324, L#338 ? That will also help with removing the subsequent `else` blocks

File: test/integration/metrics/metrics_test.go [hunk 3]
```diff
@@ -261,6 +263,90 @@ func TestAPIServerMetricsLabels(t *testing.T) {
 	}
 }
 
+func TestAPIServerMetricsLabelsWithAllowList(t *testing.T) {
+
+	allowMetricLabels := []struct {
```
Comment (richabanker) [inline test/integration/metrics/metrics_test.go:268]: This setup almost feels like a table test, so can we just convert it to that (similar to the setup of `TestAPIServerMetricsPods()`)?

Comment (yongruilin) [reply test/integration/metrics/metrics_test.go]: Updated.

File: test/integration/metrics/metrics_test.go (hunk 2 above)
Comment (yongruilin) [reply test/integration/metrics/metrics_test.go]: Updated. Thanks!

File: test/integration/metrics/metrics_test.go (hunk 1 above)
Comment (yongruilin) [reply test/integration/metrics/metrics_test.go:296]: It's an one-off operation dedicated for this test (setup testAPIServer once with the flag including all the test cases), I don't it is worth to extract this logic out.

File: test/integration/metrics/metrics_test.go [hunk 4]
```diff
@@ -261,6 +265,101 @@ func TestAPIServerMetricsLabels(t *testing.T) {
...
+			allowValuesStr = append(allowValuesStr, string(allowValue))
+		}
+		allowMetricLabelFlagStrs = append(allowMetricLabelFlagStrs, fmt.Sprintf("\"%s,%s=%s\"", tc.metricName, tc.labelName, strings.Join(allowValuesStr, ",")))
+	}
+	allowMetricLabelsFlag := "--allow-metric-labels=" + strings.Join(allowMetricLabelFlagStrs, ",")
+
+	testServerFlags := framework.DefaultTestServerFlags()
+	testServerFlags = append(testServerFlags, allowMetricLabelsFlag)
+
+	metrics.Reset()
```
Comment (richabanker) [inline test/integration/metrics/metrics_test.go:312]: why are the reset() being called twice?

Comment (richabanker) [reply test/integration/metrics/metrics_test.go:312]: Ah I see reset() is not called for the other tests. Maybe we should add a cleanup() for all tests that can do this resetting. Can have a TODO for it.

Comment (yongruilin) [reply test/integration/metrics/metrics_test.go:312]: I'm also thinking whether we should add the metrics.Reset() in the [tearDown()](https://github.com/kubernetes/kubernetes/blob/b240c51a725d6c43221709c9b7a04f937757af37/cmd/kube-apiserver/app/testing/testserver.go#L174) of the TestAPIServer.

Comment (yongruilin) [reply test/integration/metrics/metrics_test.go:313]: I've added a comment with TODO to follow up.

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve

**Comment by yongruilin** [comment]
pull-kubernetes-unit failed on due to `TestUpdateNodeStatusWithLease`
Fix https://github.com/kubernetes/kubernetes/pull/128629 has been merged need to rebase.

**Pull request merged**